packages/codemirror-blocks/src/ui/PrimitiveList.tsx

```diff
@@ -93,69 +93,139 @@ export const PrimitiveGroup = (props: PrimitiveGroupProps) => {
-      key={p.name}
-      primitive={p}
-      onFocus={onFocus}
-      onKeyDown={onKeyDown}
-      className={selected == p.name ? "selected" : ""}
-    />
-  );
+export const PrimitiveList = React.forwardRef(
+  (props: PrimitiveListProps, ref: React.MutableRefObject<HTMLElement>) => {
+    const {
+      primitives = [],
+      selected,
+      setSelectedPrimitive,
+      selectedPrimitive,
+      toolbarRef,
+      searchString,
+    } = props;
 
-  const text = searchString
-    ? (primitives.length == 0 ? "No" : primitives.length) + " blocks found"
-    : "blocks";
+    // @pcardune - typing this array has been a challenge
+    // JSX.Element doesn't work, even though that's the
+    // recommendation I'm seeing online
+    const primitiveRefs = useRef<any[]>([]);
 
-  return (
-    <div>
-      <h3
-        id="toolbar_heading"
-        className="screenreader-only"
-        aria-live="assertive"
-        aria-atomic="true"
-      >
-        {text}
-      </h3>
-      <ul
-        className="PrimitiveList list-group"
-        aria-labelledby="toolbar_heading"
-      >
-        {primitives.map((item) =>
-          item instanceof PrimitiveGroupModel
-            ? renderGroup(item)
-            : renderPrimitive(item)
-        )}
-      </ul>
-    </div>
-  );
-};
+    const renderGroup = (g: PrimitiveGroupModel) => (
+      <PrimitiveGroup
+        key={g.name}
+        group={g}
+        selected={selected}
+        selectedPrimitive={selectedPrimitive}
+        toolbarRef={toolbarRef}
+        setSelectedPrimitive={setSelectedPrimitive}
+      />
+    );
+
+    const renderPrimitive = (p: LanguagePrimitive, i: number) => {
+      const primRef = React.createRef<HTMLElement>();
+      const prim = (
+        <Primitive
+          key={p.name}
+          primitive={p}
+          onFocus={() => setSelectedPrimitive(p)}
+          className={selected == p.name ? "selected" : ""}
+          ref={primRef}
+        />
+      );
+      primitiveRefs.current[i] = prim;
+      return prim;
+    };
+
+    // Set selectedPrimitive state, depending on whether we go up or down
+    const move = (event: React.KeyboardEvent, dir: "Up" | "Down") => {
+      if (!selectedPrimitive || primitives.length == 0) {
+        return; // Nothing to select. Bail.
+      }
+
+      // compute the index of the newly-selected primitive, and get the Element
+      let newIndex;
+      const prevIndex = primitives.indexOf(selectedPrimitive); // -1 if nothing selected
+      if (dir == "Down") {
+        newIndex = Math.min(prevIndex + 1, primitives.length - 1);
+      } else {
+        newIndex = Math.max(prevIndex - 1, 0);
+      }
+      const newPrimitiveElt = primitiveRefs.current[newIndex];
+
+      // @pcardune - is it safe to assume this will always exist, and ditch the check?
+      if (newPrimitiveElt) {
+        newPrimitiveElt.ref.current.focus();
+      } else {
+        return err("No DOM node was created for " + primitives[newIndex].name);
```

Comment:
Did you mean to throw? The `err()` function just returns a typed result object. It's only useful if the object gets used somehow up the stack. Since nothing ever checks the return value of `move()`, this line is equivalent to just `return;`.